GOVERNANCE_DOC.md

```diff
@@ -0,0 +1,597 @@
+
+### Validator’s punishment for non-voting
+
+Validators are required to vote on all proposals to ensure that results have legitimacy. Voting is part of validators' directives and failure to do it will result in a penalty. 
+
+If a validator’s address is not in the list of addresses that voted on a proposal and if the vote is closed (i.e. `MinDeposit` was reached and `Voting period` is over), then this validator will automatically be partially slashed of `GovernancePenalty`.
+
+*Note: Need to define values for `GovernancePenalty`*
+
+**Exception:** If a proposal is a `Urgent` proposal and is accepted via the special condition of having a ratio of `Yes` votes to `InitTotalVotingPower` that exceeds 2/3, validators cannot be punished for not having voted on it. That is because the proposal will close as soon as the ratio exceeds 2/3, making it mechanically impossible for some validators to vote on it.
+
+### Governance key and governance address
+
+Validators can make use of an additional slot where they can designate a `Governance PubKey`. By default, a validator's `Governance PubKey` will be the same as its main PubKey. Validators can change this `Governance PubKey` by sending a `Change Governance PubKey` transaction signed by their main `Consensus PubKey`. From there, they will be able to sign vote using the `Governance PrivKey` associated with their `Governance PubKey`. The `Governance PubKey` can be changed at any moment.
+
+
+## Software Upgrade
+
+If proposals are of type `SoftwareUpgradeProposal`, then nodes need to upgrade their software to the new version that was voted. This process is divided in two steps.
+
+### Signal
+
+After a `SoftwareUpgradeProposal` is accepted, validators are expected to download and install the new version of the software while continuing to run the previous version. Once a validator has downloaded and installed the upgrade, it will start signaling to the network that it is ready to switch by including the proposal's `proposalID` in its *precommits*.(*Note: Confirmation that we want it in the precommit?*)
+
+Note: There is only one signal slot per *precommit*. If several `SoftwareUpgradeProposals` are accepted in a short timeframe, a pipeline will form and they will be implemented one after the other in the order that they were accepted.
+
+### Switch
+
+Once a block contains more than 2/3rd *precommits* where a common `SoftwareUpgradeProposal` is signaled, all the nodes (including validator nodes, non-validating full nodes and light-nodes) are expected to switch to the new version of the software. 
+
+*Note: Not clear how the flip is handled programatically*
```

Comment:
I also think we should maybe introduce a punishment for validators who do not switch within a certain period of time once governance has accepted.... this will help ease the above scenario too!

Reply:
They're already punished for being offline!

Reply:
But maybe they don't go offline if they don't upgrade, they're just always be submitting invalid blocks or voting against an upgraded proposers blocks

Reply:
But more than 2/3rd validators upgraded so the chain won't halt. And the one who did not upgraded will fail to propose and/or have their precommit missing from the blocks that get committed, and will thus be punished

Reply:
not true, they could still propose an invalid block if they haven't upgraded their protocol... I'm not sure there is an explicit way to prevent this here (maybe?) @ebuchman

Reply:
But this block will not be accepted because more than 2/3rd validators think it's invalid. So they will not propose valid blocks until they upgrade, and their precommit will not be inlcuded until they upgrade -> they will be punished

Reply:
If a block contains 2/3rd precommits for a software upgrade, then the next proposal should be with the new software version, i.e., the proposal is considered invalid if it does not have the new version. This require having version as an additional field in the message, but we should probably have this for various reasons.

Reply:
Opened an issue in the SDK for this, probably needs to be a solution across SDK/Tendermint https://github.com/cosmos/cosmos-sdk/issues/403